test/gtest/ucp/test_ucp_tag.cc

```diff
@@ -423,22 +424,116 @@ class test_ucp_tag_limits : public test_ucp_tag {
+
+        add_variant_with_value(variants, params, 0, "");
+        add_variant_with_value(variants, params, 1, "offload");
     }
 
 protected:
     bool m_test_offload;
+
+    void check_rndv_startup_config(size_t exp_rndv_thresh,
+                                   size_t exp_rndv_intra_thresh,
+                                   size_t exp_rndv_inter_thresh)
+    {
+        ucp_context_config_t *cfg = &sender().worker()->context->config.ext;
+
+        EXPECT_EQ(exp_rndv_thresh, cfg->rndv_thresh);
+        EXPECT_EQ(exp_rndv_intra_thresh, cfg->rndv_intra_thresh);
+        EXPECT_EQ(exp_rndv_inter_thresh, cfg->rndv_inter_thresh);
+    }
+
+    void check_tag_rndv(size_t cfg_thresh)
+    {
+        ucp_ep_config_t *cfg = ucp_ep_config(sender().ep());
+
+        if (m_test_offload) {
+            /* There is a lower bound for rndv threshold with tag offload.
+             * We should not send messages smaller than this limit */
+            size_t min_rndv = ucp_ep_tag_offload_min_rndv_thresh(sender().ucph(),
+                                                                 &cfg->key);
+
+            /* If configured threshold is less than min_rndv, then expect exact
+             * min_rndv limit for short messages */
+            if (cfg_thresh < min_rndv) {
+                EXPECT_EQ(cfg->tag.offload.max_eager_short.memtype_on + 1,
+                          min_rndv);
+                EXPECT_EQ(cfg->tag.offload.max_eager_short.memtype_off + 1,
+                          min_rndv);
+            } else {
+                EXPECT_LE(cfg->tag.offload.max_eager_short.memtype_on + 1,
+                          cfg_thresh);
+                EXPECT_LE(cfg->tag.offload.max_eager_short.memtype_off + 1,
+                          cfg_thresh);
+            }
+        } else {
+            EXPECT_LE(cfg->tag.max_eager_short.memtype_on + 1, cfg_thresh);
+            EXPECT_LE(cfg->tag.max_eager_short.memtype_off + 1, cfg_thresh);
+        }
+    }
+
+    void check_am_rndv(size_t cfg_thresh)
+    {
+        ucp_ep_config_t *cfg = ucp_ep_config(sender().ep());
+
+        EXPECT_LE(cfg->am_u.max_eager_short.memtype_on + 1, cfg_thresh);
+        EXPECT_LE(cfg->am_u.max_eager_short.memtype_off + 1, cfg_thresh);
+        EXPECT_LE(cfg->am_u.max_reply_eager_short.memtype_on + 1, cfg_thresh);
+        EXPECT_LE(cfg->am_u.max_reply_eager_short.memtype_off + 1, cfg_thresh);
+    }
+
+    void check_ep_proto_rndv(size_t cfg_thresh)
+    {
+        ucp_ep_config_t *cfg = ucp_ep_config(sender().ep());
+        ucp_proto_select_t *proto_select = &cfg->proto_select;
+        ucp_proto_select_elem_t value;
+        size_t msg_length = cfg_thresh;
+
+        /* Check tag matching thresholds */
+        if (m_test_offload) {
+            /* There is a lower bound for rndv threshold with tag offload.
+             * We should not send messages smaller than this limit */
+            size_t min_rndv = ucp_ep_tag_offload_min_rndv_thresh(sender().ucph(),
```

Comment:
minor: maybe it can be done once at init stage and saved in some member variable

Reply:
right, good idea indeed